### ISSUE TITLE
Convert discovery processors into entity providers

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#10183]
## Feature Suggestion

The following discovery processors would benefit from being converter to using the `EntityProvider` interface instead, as among other things that will enable immediate deletions rather than orphaned entities, as well as more granular control of the update interval. More details about entity providers can be found in the [catalog customization docs](https://backstage.io/docs/features/software-catalog/external-integrations#custom-entity-providers).

- [ ] [AwsS3DiscoveryProcessor](https://github.com/backstage/backstage/blob/66ba5d9023805083519ac1b3df95840039d0fa77/plugins/catalog-backend-module-aws/src/processors/AwsS3DiscoveryProcessor.ts)
- [ ] [GithubDiscoveryProcessor](https://github.com/backstage/backstage/blob/66ba5d9023805083519ac1b3df95840039d0fa77/plugins/catalog-backend-module-github/src/GithubDiscoveryProcessor.ts)
- [ ] [GitLabDiscoveryProcessor](https://github.com/backstage/backstage/blob/66ba5d9023805083519ac1b3df95840039d0fa77/plugins/catalog-backend-module-gitlab/src/GitLabDiscoveryProcessor.ts)
- [ ] [BitbucketDiscoveryProcessor](https://github.com/backstage/backstage/blob/66ba5d9023805083519ac1b3df95840039d0fa77/plugins/catalog-backend-module-bitbucket/src/BitbucketDiscoveryProcessor.ts)
- [ ] [AzureDevOpsDiscoveryProcessor](https://github.com/backstage/backstage/blob/66ba5d9023805083519ac1b3df95840039d0fa77/plugins/catalog-backend-module-azure/src/AzureDevOpsDiscoveryProcessor.ts)

**Comment by pjungermann** [comment]
I'm interested in awsS3, bitbucket, and github. I didn't start working on any of these yet, but I'm open to contributing. If anyone starts to work on these, would be great to share it here (to prevent double work).

**Comment by pjungermann** [comment]
@Rugvip as an MVP would it be enough if the providers do the same as the discovery processors and create Location entities which then get picked up by the other processors? Or might that still cause orphans?

**Comment by Rugvip** [comment]
@pjungermann yeah that should be alright, probably the way we wanna implement it tbh. Eager deletions happen recursively so if a deleted location is the only thing referencing an entity then the entity will be deleted as well

**Comment by pjungermann** [comment]
I will look at replacing/converting the BitbucketDiscoveryProcessor next. Hereby, I plan to consider #9923 and create separate providers for Bitbucket Cloud and Bitbucket Server (fka. Stash).

**Comment by arslan-raza-143** [comment]
[https://github.com/bitcoin/bitcoin/issues/24783#issue-1194493518](https://github.com/bitcoin/bitcoin/issues/24783#issue-1194493518) @pjungermann

**Comment by brentg-telus** [comment]
I'm interested in taking on the GithubDiscoveryProcessor and converting it to an Entity Provider.

**Comment by goober** [comment]
@pjungermann Have you started with converting the BitbucketServerEntityProvider yet? Otherwise I can grab that one.

**Comment by pjungermann** [comment]
@goober No, I'm working on Bitbucket Cloud at the moment. I'm nearly done, need to polish a few things. I plan to create a PR soon. I decided to put this into a new module `plugins/catalog-backend-module-bitbucket-cloud`(may change still) instead of `plugins/catalog-backend-module-bitbucket` (I want to have both separated so you can pull in the one(s) you actually need/use).

**Comment by goober** [comment]
> @goober No, I'm working on Bitbucket Cloud at the moment. I'm nearly done, need to polish a few things. I plan to create a PR soon. I decided to put this into a new module `plugins/catalog-backend-module-bitbucket-cloud`(may change still) instead of `plugins/catalog-backend-module-bitbucket` (I want to have both separated so you can pull in the one(s) you actually need/use).

Thanks. Then I will wait until you have contributed your changes and will then finalize the BitbucketServerEntityProvider

**Comment by goenning** [comment]
The PR above is for the Azure DevOps implementation

**Comment by pjungermann** [comment]
PR #12334 will contribute the provider for GitHub

**Comment by pjungermann** [comment]
@goober are you still up for taking on Bitbucket Server?

**Comment by goober** [comment]
> @goober are you still up for taking on Bitbucket Server?

I moved on to another employer so unfortunately not. However, we got it working internally before I left, so maybe @ONordander wants to take a look at a contribution?

**Comment by pjungermann** [comment]
@namco1992 sounds good. Feel free to reach out if you run into challenges. You can drop your questions at the support channel in Discord, too.

@ONordander in case you have something which you could share and therefore, @namco1992 would not need to invest time, please let us know.

**Comment by pjungermann** [comment]
PS: we only have Bitbucket Server and GitHub (PR under review) left still.

**Comment by namco1992** [comment]
@ONordander Cool! I will leave it for you then, look forward to your PR 😃

**Comment by ONordander** [comment]
@pjungermann looking into what we are running locally and realized that I need to make changes to https://github.com/backstage/backstage/tree/master/packages/integration/src/bitbucketServer, to add possibility for username/password authentication.
What's the easiest way of doing that, a separate PR for the integration?

**Comment by pjungermann** [comment]
@ONordander I got you covered with that via PR #12711 related to another issue. It is waiting for confirmation by a Bitbucket Server user. Currently, we were hoping that @namco1992 can test it. But maybe you can give it a shot, too :)

**Comment by ONordander** [comment]
> @ONordander I got you covered with that via PR #12711 related to another issue. It is waiting for confirmation by a Bitbucket Server user. Currently, we were hoping that @namco1992 can test it. But maybe you can give it a shot, too :)

Nice, I'll try to get it running today for testing.

**Comment by marcus-crane** [comment]
@Rugvip I think this issue is complete now? 🤔

**Comment by Rugvip** [comment]
Indeed, thanks! 😁 

And thank you everyone that helped out with this! 🙏
